feat: invite function porting from SafetyNet

exo404 opened the pull request.

exo404 committed:
feat: invite function porting from SafetyNet

exo404 committed:
fix: formatting

exo404 commented on script/InviteGenerator.sol:
```diff
@@ -0,0 +1,102 @@
+    _inviteTypeHash = keccak256(bytes(inviteTypeString));
+  }
+  /// @notice Returns the struct hash of an invite
+
```
[d257086](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/d2570860c7ad0000570dbab299bcaaa6eacc7149)

exo404 commented:
@RonTuretzky @bagelface all tests are passing I think we can already merge this

RonTuretzky commented on script/InviteGenerator.sol:
```diff
@@ -0,0 +1,102 @@
+  bytes32 private _inviteTypeHash;
+
+  /// @notice EIP-712 domain type hash
+  bytes32 private constant _EIP712_DOMAIN_TYPEHASH =
```
why not just have the bytes and the string in a comment

exo404 replied:
Do you want to delete all the constants and put them in the comment? @RonTuretzky

RonTuretzky replied:
yes best practice whe hashing strings is to just store the hash and put in a comment above the hash what it is ,there's no benefit to storing the string in the contract

exo404 replied:
[262870d](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/262870dad510e56932891d12cf47415518a839d9)

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -19,12 +20,32 @@ import {ISavingCircles} from 'interfaces/ISavingCircles.sol';
+  string private constant _INVITE_SIGNATURE_VERSION = '1';
+
+  /// @notice EIP-712 type hash for invite signatures
+  bytes32 private constant _INVITE_TYPEHASH = keccak256('Invite(uint256 circleId,uint256 nonce)');
```
have bytes and string in comment

exo404 replied:
[37d1cb9](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/37d1cb9745c9bded71bbb5a567edb92c45c66ab5)

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -19,12 +20,32 @@ import {ISavingCircles} from 'interfaces/ISavingCircles.sol';
   uint256 public constant MINIMUM_MEMBERS = 2;
 
+  /// @notice Invite signing domain name used for EIP-712 signatures
+  string private constant _INVITE_SIGNING_DOMAIN = 'StacksInvite';
```
have bytes and string in comment

exo404 replied:
[37d1cb9](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/37d1cb9745c9bded71bbb5a567edb92c45c66ab5)

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -19,12 +20,32 @@ import {ISavingCircles} from 'interfaces/ISavingCircles.sol';
+  bytes32 private constant _INVITE_TYPEHASH = keccak256('Invite(uint256 circleId,uint256 nonce)');
+
+  /// @notice EIP-712 domain type hash
+  bytes32 private constant _EIP712_DOMAIN_TYPEHASH =
```
have bytes and string in comment

exo404 replied:
[37d1cb9](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/37d1cb9745c9bded71bbb5a567edb92c45c66ab5)

RonTuretzky commented on src/interfaces/ISavingCircles.sol:
```diff
@@ -24,6 +24,14 @@ interface ISavingCircles {
+  /// @notice Struct defining an invite to join a circle
+  /// @param circleId ID of the Circle
+  /// @param nonce Unique nonce for the invite
+  struct Invite {
```
seems like a redundant struct imo

exo404 replied:
why? I can take the circleId and nonce from other sources but this seems easier to read in the code

RonTuretzky replied:
because it's just two ints

exo404 replied:
Yes I agree but this is the whole purpose of using EIP712 standard. It works with structs so we can extend the invite type in the future. I was deleting it but then I thought this

exo404 replied:
Ref [here](https://eips.ethereum.org/EIPS/eip-712)

RonTuretzky replied:
can you reference this? also can you provide another reference that is doing what you are doing?  [this](https://github.com/Layr-Labs/eigenlayer-contracts/blob/57c37ae5829a1d2abb42fcae05774201ac3716fe/src/contracts/permissions/KeyRegistrar.sol#L344) impl seems to be different

RonTuretzky replied:
@exo404 can you ref specifically the part in the spec supports changing the struct?

exo404 replied:
@RonTuretzky there's a signing version parameter

exo404 replied:
<img width="2109" height="654" alt="image" src="https://github.com/user-attachments/assets/87db0928-f6a2-4728-88ea-681219c16104" />

exo404 replied:
So I think that we can change the struct and so the domain version if we need it

RonTuretzky replied:
@exo404 did you happen to research to existing invitation link solidity implementations?

exo404 replied:
> can you reference this? also can you provide another reference that is doing what you are doing? [this](https://github.com/Layr-Labs/eigenlayer-contracts/blob/57c37ae5829a1d2abb42fcae05774201ac3716fe/src/contracts/permissions/KeyRegistrar.sol#L344) impl seems to be different

you have to look at their '_calculateSignableDigest' function, that's the same function as our _hashInvite

exo404 replied:
I believe that the structHash depends on what we are signing with EIP712, correct me if I'm wrong

exo404 replied:
> @exo404 did you happen to research to existing invitation link solidity implementations?

I simply tried to follow the standard as I said last meeting

RonTuretzky replied:
I'm starting to think that was a mistake and I should have looked for a reference

exo404 replied:
@bagelface I need a review here

bagelface replied:
I think the "struct Invite" definition here is redundant because we are using variables that are already defined in the contract. Instead, we can implement the hashing function like this and still have a valid 712 implementation:

```solidity
function _hashInvite(uint256 _circleId, uint256 _nonce) private view returns (bytes32) {
    bytes32 _inviteTypehash = keccak256('Invite(uint256 circleId,uint256 nonce)');
    bytes32 _structHash = keccak256(abi.encode(_inviteTypehash, _circleId, _nonce));
    bytes32 _eip712DomainTypehash =
      keccak256('EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)');
    ...
  }
```

exo404 replied:
Ok perfect, I changed it deleting the struct. Test are still passing [34f2310](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/34f23100e8f86ce7b7817854ea9f6ec53b88a315)

RonTuretzky reviewed (changes requested):
please add frontend documentation and compat snippet

exo404 commented:
> please add frontend documentation and compat snippet

I'm working on it in a different PR @RonTuretzky

exo404 committed:
fix: constants moved to the hash function

exo404 committed:
fix: hash constant

exo404 commented:
> please add frontend documentation and compat snippet

btw there's this example in the EIP712 spec [here](https://[eips.ethereum.org/assets/eip-712/Example.js](https://eips.ethereum.org/assets/eip-712/Example.js))

exo404 committed:
fix: delete Invite struct

exo404 commented:
@RonTuretzky can you review here so I can merge this and finalize the frontend component directly on stacks?

RonTuretzky commented:
@exo404  i'm not merging this until i see the frontend componenet 🙃

RonTuretzky commented on script/InviteGenerator.sol:
```diff
@@ -26,12 +26,12 @@ contract InviteGenerator is Script {
   error InvalidSignatureVersion();
 
-  constructor(string memory inviteSigningDomain, string memory inviteSignatureVersion, string memory structName) {
+  constructor(string memory inviteSigningDomain, string memory inviteSignatureVersion) {
```
small nit is that i don't understand why you keep insisting to have the strings on chain 😵

exo404 replied:
I'm sorry I didn't understand the issue, I'm changing it asap

exo404 replied:
[0b7ee91](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/0b7ee9118f80ed2545501eee3323a98bfd579767) @RonTuretzky

exo404 committed:
feat: frontend components

exo404 commented:
@RonTuretzky [e894b0c](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/e894b0c4ae90c909cdcf4d6486622d7fdb86be7a) this for frontend components

exo404 committed:
fix: delete constant strings

exo404 commented:
@bagelface I updated the code with 
- [34f2310](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/34f23100e8f86ce7b7817854ea9f6ec53b88a315)
- [0b7ee91](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/0b7ee9118f80ed2545501eee3323a98bfd579767)
- [e894b0c](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/e894b0c4ae90c909cdcf4d6486622d7fdb86be7a)

the last one is for frontend components

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -137,6 +139,27 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
   }
 
+  /// @inheritdoc ISavingCircles
+  function redeemInvite(uint256 _id, uint256 _nonce, bytes calldata _signature) external override nonReentrant {
```
@exo404 currently we're not checking we've reached the max number of invites and we're also not checking that enough members have joined before circle start time

exo404 replied:
@RonTuretzky I don't see a maximum number of members, am I missing something? Also where do you want to check the minimum members constant? Do you want to check it inside the redeem function?

RonTuretzky replied:
the circle struct has a members array. this inherently implies that there is a max member count, as we init the circle with the member array. however this pr breaks this functionality, because we no longer have ot initalize the members array. this breaks the whole contract becuase the previous pattern basically had an assured static member size count when it was initalized, and this one does not. I am unsure that this contract design supports dyanmically joining , as it was never intended to be used like this. does that make sense?

exo404 replied:
I didn't think about this but you are 100% right. I remember the contract assumed this from the start. Ho do we proceed? I can try to refactor the contract to include the dynamic members number

RonTuretzky replied:
you can start with some tests that check what happens through this behavior to understand if we've broken the contract or not. the safest approach would be to remove the members area from the struct,add a counter that increments for each invite acceptance, and  add in the decomission check to see if enough members have joined so the circle is locked until everyone has joined

franrolotti committed:
feat: hard-deprecate maxDeposits in favor of members.length and align contracts/tests

franrolotti committed:
feat: introduce circleEnd field and use stored value across contracts; guard overflow in create()

exo404 committed:
Merge pull request #62 from BreadchainCoop/54-assess-whether-to-deprecate-maxdeposits-in-favor-of-memberslength

merge #54 deprecate maxdeposits in #51 invite function

franrolotti committed:
test: fix fuzz expectation by matching depositAmount-first revert order

exo404 committed:
feat: refactor to invite new members

exo404 committed:
fix: member check in create function

exo404 committed:
test: refactor tests for the new user flow

exo404 committed:
fix: add check to redeemInvite for active circles

exo404 committed:
fix: checks and start time

exo404 committed:
test: fix fuzz

exo404 committed:
test: fix integration

exo404 committed:
test: fix invariant

exo404 committed:
test: fix unit

exo404 committed:
fix: add onlymember to withdraw

exo404 committed:
test: fix unit

exo404 commented:
@RonTuretzky @franrolotti the invite feature is implemented, also with @diterra-code request for multiple invites generation. I fixed all the tests. Let me know if there's something to change. While you review this I'm focusing on Gelato's issue

exo404 committed:
Merge branch 'dev' into 51-invite-function-integration-from-safety-net

exo404 committed:
Revert "Merge branch 'dev' into 51-invite-function-integration-from-safety-net"

This reverts commit b7d76bee501766e53d4c821857ac8c6ab5245cc8, reversing
changes made to f12bfed49ec5e4c64dda24dd9d98047e1b7559f8.

RonTuretzky commented on frontend/GenerateInviteForm.tsx:
```diff
@@ -0,0 +1,173 @@
+        const circleId = BigInt(circleIdInput);
+        const nonce = BigInt(nonceInput);
+        const verifyingContract = verifyingContractInput.trim();
+        const privateKey = privateKeyInput.trim();
```
we cannot take the users private key in the frontend as a raw value, here is a reference for how to sign using a privy wallet integration https://github.com/communetxyz/sharehouse-app/blob/e92297e71da19818ac65983ea0bd88d1948d817e/app/dashboard/invites/page.tsx#L75

RonTuretzky replied:
please investigate if it's possible to authorize multiple wallet signatures with one click

exo404 committed:
fix: revert changes to use Ron reference

exo404 committed:
feat: invite page

exo404 committed:
test: invite page test

exo404 committed:
chore: add dev dependencies fro react developement

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -25,6 +26,8 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
   mapping(uint256 id => mapping(address member => bool status)) public isMember;
   mapping(address member => uint256[] ids) public memberCircles;
   mapping(address token => bool status) public allowedTokens;
+  mapping(uint256 id => mapping(uint256 nonce => bool used)) public usedNonces;
```
Shouldn't this be per circle per user?

franrolotti commented on app/invites.tsx:
```diff
@@ -0,0 +1,569 @@
+import { BrowserProvider, Contract } from 'ethers';
+
+const PRIVY_APP_ID = (import.meta as any).env?.VITE_PRIVY_APP_ID ?? 'privy-app-id';
+const DEFAULT_SAVING_CIRCLES_ADDRESS =
```
We should use a fixed app config for the contract address

franrolotti commented on app/invites.tsx:
```diff
@@ -0,0 +1,569 @@
+  const { wallets, ready: walletsReady } = useWallets();
+
+  const [contractAddress, setContractAddress] = useState(DEFAULT_SAVING_CIRCLES_ADDRESS);
+  const [circleId, setCircleId] = useState('');
```
like the commune flow, invites should be scoped to a specific circle. the generator of the invites has to be the owner of the circle. Might be nicer to have `circleId` come from route params / app state rather than user input

franrolotti commented on app/invites.tsx:
```diff
@@ -0,0 +1,569 @@
+      const signer = await getSignerFromPrivyWallet(activeWallet);
+      const network = await signer.provider?.getNetwork();
+      const chainId = network?.chainId ?? 0n;
+      const contract = new Contract(contractAddress, savingCirclesAbi, signer);
```
As we want only the circle owner to create invite links (like the commune invite flow). Since the ABI already exposes `owner()`, could we add an explicit owner check here?

And ideally add a “You’re not the owner” UI state instead of just a generic status.

franrolotti commented on app/invites.tsx:
```diff
@@ -0,0 +1,569 @@
+        const alreadyUsed = await contract.usedNonces(parsedCircleId, candidate);
+        if (!alreadyUsed) {
+          const typedData = buildInviteTypedData(parsedCircleId, candidate, chainId, contractAddress);
+          const signature = await signInvite(signer, typedData);
```
This option doesn't batch all invites into a single signing step, nor hides the multiple signatures as we spoke in the meeting. We should solve this here

franrolotti commented on app/invites.tsx:
```diff
@@ -0,0 +1,569 @@
+      <section style={{ marginBottom: 22 }}>
+        <h3 style={{ margin: '0 0 8px 0' }}>Redeem an invite</h3>
+        <p style={{ margin: '0 0 10px 0', color: '#475569' }}>
+          Paste a signed invite to redeem it for the connected Privy wallet. This calls SavingCircles.redeemInvite.
```
May be better for the user to click an invite URL and the page auto-redeems those params for them (with a confirmation UI), without them having to copy/paste anything.

franrolotti reviewed (commented):
Right now this page feels more like an internal dev/admin tool than a product-facing “invite to a circle” flow frontend. I think the goal is to be closer to the commune invite UX.

We should refactor to keep the to keep all the EIP-712 + `usedNonces` logic but wrap it in a polished, app-styled UI like the commune invite page. 

Also we should refactor to

* Only circle owner can create invites.
* The page should be scoped to a specific circle where users don't type `circleId` or `contractAddress`
* Right now, multiple invites trigger multiple signature popups
* Use design system for the UI


My suggestion is to keep all the strong crypto/contract work from this PR, but refactor the component so this is the final Saving Circles invite experience (owner-only, single-circle, one-click create + URL-based redeem, commune-style UI), rather than a generic admin-style tool.

@RonTuretzky wdyt

RonTuretzky commented:
> Right now this page feels more like an internal dev/admin tool than a product-facing “invite to a circle” flow frontend. I think the goal is to be closer to the commune invite UX.
> 
> We should refactor to keep the to keep all the EIP-712 + `usedNonces` logic but wrap it in a polished, app-styled UI like the commune invite page.
> 
> Also we should refactor to
> 
> * Only circle owner can create invites.
> * The page should be scoped to a specific circle where users don't type `circleId` or `contractAddress`
> * Right now, multiple invites trigger multiple signature popups
> * Use design system for the UI
> 
> My suggestion is to keep all the strong crypto/contract work from this PR, but refactor the component so this is the final Saving Circles invite experience (owner-only, single-circle, one-click create + URL-based redeem, commune-style UI), rather than a generic admin-style tool.
> 
> @RonTuretzky wdyt

It should not be product facing, its just a reference for the frontend ppl to be able to have a working reference so a really janky page is fine as long as it is functional

exo404 committed:
fix: addressing francisco comments

exo404 committed:
fix: issue #64

exo404 commented:
[5317da8](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/5317da8b21bea9e3746d7307dca46cb5d5d56d3a) fixes #64

exo404 committed:
fix: issue #66

exo404 commented:
[1d466d5](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/1d466d5cb03cae2546e212de57c85647248955f5) fixes ##66

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -94,7 +87,25 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
     if (msg.sender != _circle.owner) revert NotOwner();
-    if (_circle.members.length < MINIMUM_MEMBERS) revert InvalidMemberCount();
+    uint256 membersCount = 0;
+    for (uint256 i = 0; i < _circle.members.length; i++) {
```
do we really need two loops here or am I missing something?

exo404 replied:
I think we need to know the members array dimension before declaring it right? I doesn't work with dynamic arrays

RonTuretzky replied:
https://github.com/BreadchainCoop/saving-circles/pull/70

exo404 replied:
I leaved a comment there @RonTuretzky

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -198,26 +262,28 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
   /**
    * @dev Make a withdrawal from a specified circle
-   *      Permissionless: anyone can trigger the payout for the member whose turn it is to withdraw
+   *      A withdrawal must be made by a member of the circle, even if it is for another member.
```
when did we decide on this change?

exo404 replied:
I'm not sure I made this change tbh, but I can revert this. I see there's only one check in the function for this

exo404 replied:
[72f0c2c](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/72f0c2cac3f6bc2b4b1363cb599b577efd30d1c3)

exo404 committed:
fix: revert changes to withdraw function

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -291,9 +291,9 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
 
-    _circle.currentIndex = (_circle.currentIndex + 1) % _circle.members.length;
-    bool success = IERC20(_circle.token).transfer(_member, _withdrawAmount);
+    bool success = IERC20(_circle.token).transfer(_circle.members[_circle.currentIndex], _withdrawAmount);
```
i don't understand this change , why is this in this function? please decouple it so that this is determined in the func that is calling it (so withdraw for should determine whether the perosn is eligible for a withdrawl. also , i think it could be they're not the current index but can still withdraw? unsure?

exo404 replied:
[60abb0d](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/60abb0d6eceab448f8f113c8fb8cba982e7d7a0b)

RonTuretzky committed:
fix: optimize start() to use single loop for member filtering

exo404 committed:
fix: comments for the withdraw function

exo404 commented on src/contracts/SavingCircles.sol:
```diff
@@ -63,48 +66,84 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+    circles[_id].members = members;
 
-    emit CircleCreated(_id, _circle.members, _circle.token, _circle.depositAmount, _circle.depositInterval);
+    _circle.circleStart = block.timestamp;
```
we want to override it because we need to calculate the correct indexes

RonTuretzky replied:
i think there's a difference between "circleMinimumStart" and "circleStart" which is basically what we're missing here otherwise its confusing.

exo404 replied:
what do you mean?

RonTuretzky replied:
There is a difference between "don't start before this time" and "this started at this time"

RonTuretzky replied:
There is a difference between "don't start before this time" and "this started at this time"

exo404 replied:
ok I see but we have got only the second one. Also does make sense to have a "don't start before this time" variable?

RonTuretzky replied:
What are we initializing then in the create? Which one is it supposed to be? If it's not supposed to be "don't start before this" then the create should not be initializing it

RonTuretzky replied:
What are we initializing then in the create? Which one is it supposed to be? If it's not supposed to be "don't start before this" then the create should not be initializing it

exo404 replied:
We actually don't touch the circleStart field in the create function, but we are passing the circle struct containing a circleStart value, that's it

RonTuretzky replied:
ok this is what is confusing copilot , and its correct in the sense that the naming is ambigious. lets rename it to "effectiveCircleStartTime" which is only set when the circle actually starts, and make sure during circle creation its 0. we can also use this instead of the isActive mapping, but I kind of want to push this PR through and not refactor anyfurther

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -63,48 +66,84 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+      address member = _circle.members[i];
+      if (member == address(0)) {
+        continue;
+      }
+      members[memberIndex] = member;
+      memberIndex++;
+    }
+    circles[_id].members = members;
 
-    emit CircleCreated(_id, _circle.members, _circle.token, _circle.depositAmount, _circle.depositInterval);
+    _circle.circleStart = block.timestamp;
+    uint256 len = _circle.members.length;
+    uint256 maxDelta = type(uint256).max - _circle.circleStart;
+    if (_circle.depositInterval > maxDelta / len) revert InvalidDepositInterval();
+    _circle.circleEnd = _circle.circleStart + (_circle.depositInterval * len);
 
-    return _id;
+    isActive[_id] = true;
+    emit CircleStarted(_id);
   }
```
@exo404 i agree

exo404 replied:
yes me too, I'll add tests for this

exo404 replied:
[436855f](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/436855ff7515b32b7e5e5208c477fe3a30ea876b)

exo404 commented on src/contracts/SavingCircles.sol:
```diff
@@ -63,48 +66,84 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+      memberIndex++;
+    }
+    circles[_id].members = members;
 
```
I think I need to delete the check in the invite function

exo404 replied:
Because the frontend will pass circle struct with less members than the actual array lenght

exo404 replied:
@RonTuretzky  wdyt?

RonTuretzky replied:
I don't understand please rewrite more clearly , which check, when is the frontend passing in what, how is this related to the comment

RonTuretzky replied:
I don't understand please rewrite more clearly , which check, when is the frontend passing in what, how is this related to the comment

exo404 replied:
Ok so first of all, I don't think ordering is important because we are not considering it. So the first part of the Copilot comment is useless. The second thing is that currently we pass a members array with the maximum size specified in the frontend, but without all the members addresses. So the circle passed to the create function got a members array with some address(0) in it. Maybe we can change the struct to take a new field called size to avoid this

exo404 replied:
@RonTuretzky

exo404 replied:
[f75f632](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/f75f632c5a2792fca3fb1916779c11901495b584)

RonTuretzky replied:
why are we even doing this during the start function? why don't we update the circle member count and the members array during the redeem invite function? whats the utility of having it in this function?

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -133,10 +172,33 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
 
     emit CircleDecommissioned(_id);
   }
 
+  /// @inheritdoc ISavingCircles
+  function redeemInvite(uint256 _id, uint256 _nonce, bytes calldata _signature) external override nonReentrant {
+    Circle storage _circle = circles[_id];
+
+    if (_circle.owner == address(0)) revert NotCommissioned();
+    if (usedNonces[_id][_nonce]) revert InviteAlreadyUsed();
+    if (isMember[_id][msg.sender]) revert AlreadyMember();
+    if (isActive[_id]) revert AlreadyActive();
+
+    bytes32 _digest = _hashInvite(_id, _nonce);
+    address _signer = ECDSA.recover(_digest, _signature);
+
+    if (_signer != _circle.owner) revert InvalidSigner();
+
+    usedNonces[_id][_nonce] = true;
+    isMember[_id][msg.sender] = true;
+    memberCircles[msg.sender].push(_id);
+    _circle.members.push(msg.sender);
+
+    emit InviteRedeemed(_id, msg.sender);
+  }
```
@exo404 wdyt?

exo404 replied:
Yes I implemented tests before but I was using the old contract to check it. It was ok with that, but maybe we need new tests for it? Let me know I got no problems doing new tests

RonTuretzky replied:
Yes we need good coverage

exo404 replied:
[3e0a9ef](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/3e0a9efef701a7eb86e3837ce5f9b9230673f369)

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -211,23 +277,23 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
     if (!success) revert TransferFailed();
+    _circle.currentIndex = (_circle.currentIndex + 1) % _circle.members.length;
 
     emit FundsWithdrawn(_id, _member, _withdrawAmount);
```
@exo404 is this addressed by your revert?

exo404 replied:
[60abb0d](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/60abb0d6eceab448f8f113c8fb8cba982e7d7a0b)

exo404 commented on src/contracts/SavingCircles.sol:
```diff
@@ -211,23 +277,23 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+   *      A withdrawal must be made by a member of the circle, even if it is for another member.
    */
-  function _withdraw(uint256 _id, address _member) internal {
+  function _withdraw(uint256 _id, address _member) internal onlyMember(_id, msg.sender) {
```
[60abb0d](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/60abb0d6eceab448f8f113c8fb8cba982e7d7a0b)

exo404 commented:
[check] what happens if no one withdraws?

exo404 committed:
fix: cleaning dependencies

exo404 committed:
fix: op ordering

exo404 commented on src/contracts/SavingCircles.sol:
```diff
@@ -63,48 +66,84 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+
+      isMember[_id][member] = true;
+      memberCircles[member].push(_id);
     }
```
[2824d76](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/2824d76feda6f47350961fc0ed05e3552e81bad4)

exo404 committed:
Merge pull request #70 from BreadchainCoop/fix-start-double-loop

fix: optimize start() to use single loop for member filtering

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -295,4 +361,24 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+      abi.encode(_eip712DomainTypehash, _inviteDomainNameHash, _inviteDomainVersionHash, block.chainid, address(this))
+    );
+
+    return keccak256(abi.encodePacked('\x19\x01', _domainSeparator, _structHash));
```
I think we can ignore this but let's make a backlog issue @brendaasfjnso

RonTuretzky replied:
I think we can ignore this but let's make a backlog issue @brendaasfjnso

exo404 committed:
fix: create function to accept empy arrays of members

exo404 committed:
test: start function testing

exo404 committed:
test: redeemInvite function unit testing

exo404 commented on src/contracts/SavingCircles.sol:
```diff
@@ -63,48 +66,81 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+
+    address[] memory members = new address[](membersCount);
+    for (uint256 i = 0; i < membersCount; i++) {
+      members[i] = tempMembers[i];
```
This was my old idea but I think we resolved this with Ron's PR

exo404 commented:
> [check] what happens if no one withdraws?

If no one claims the circle, no one can deposit because the claim index can't be updated but with the withdraw function. So the circle is freezed unless it expires or someone use the withdraw function @RonTuretzky 

lines 318-322 stop anyone to deposit until currentIndex is updated

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -71,9 +71,8 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
     for (uint256 i = 0; i < _circle.members.length; i++) {
       address member = _circle.members[i];
-      if (member == address(0)) revert InvalidMemberAddress();
+      if (member == address(0)) continue;
```
wait why did we delete teh revert here?

exo404 replied:
the problem with the members list is that now we got a dynamic list updated everytime we redeem a function, so a static array in the circle struct is hard to manage. I'm trying to refactor this using a mapping instead of the static array

exo404 committed:
refactor: add members mapping and rename circleStart struct field

exo404 committed:
test: fix fuzz

exo404 committed:
test: fix integration

exo404 committed:
test: fix invariant

exo404 committed:
test: fix unit

exo404 commented:
@RonTuretzky @franrolotti I added a circleMembers mapping to replace the members array. I renamed the circleStart to effectiveCircleStartTime and I added a revert if it's not set to zero at circle creation. Now we update the membersMapping everytime we redeem an invite. I updated the Viewer and Delegated contract to match these changes and fixed the tests.

RonTuretzky commented:
> @RonTuretzky @franrolotti I added a circleMembers mapping to replace the members array. I renamed the circleStart to effectiveCircleStartTime and I added a revert if it's not set to zero at circle creation. Now we update the membersMapping everytime we redeem an invite. I updated the Viewer and Delegated contract to match these changes and fixed the tests. 

Fuck another massive change... Will review tmrw

RonTuretzky commented:
> @RonTuretzky @franrolotti I added a circleMembers mapping to replace the members array. I renamed the circleStart to effectiveCircleStartTime and I added a revert if it's not set to zero at circle creation. Now we update the membersMapping everytime we redeem an invite. I updated the Viewer and Delegated contract to match these changes and fixed the tests. 

Fuck another massive change... Will review tmrw

exo404 commented:
> > @RonTuretzky @franrolotti I added a circleMembers mapping to replace the members array. I renamed the circleStart to effectiveCircleStartTime and I added a revert if it's not set to zero at circle creation. Now we update the membersMapping everytime we redeem an invite. I updated the Viewer and Delegated contract to match these changes and fixed the tests.
> 
> Fuck another massive change... Will review tmrw

this was the only option to fix this technical debt. Let me know if it's all clear

RonTuretzky commented on test/unit/SavingCirclesUnit.t.sol:
```diff
@@ -64,19 +74,111 @@ contract SavingCirclesUnit is Test {
+    bytes32 digest = keccak256(abi.encodePacked('\x19\x01', domainSeparator, structHash));
+    (uint8 v, bytes32 r, bytes32 s) = vm.sign(_signerKey, digest);
+    return abi.encodePacked(r, s, v);
   }
```
@exo404 seems important

exo404 replied:
[5201e5f](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/5201e5fbf12872dd19f8a2d163058a14d88b8569)

[bf432fe](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/bf432fe579f5f5734fcebbd813e1cc604b1bf2ea)

[ed893b6](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/ed893b6a4e8870d956e3a4c7caae35494fc1aedc)

[6aeb0d1](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/6aeb0d125668a8b3b7f18c21c65a3248b09f6533)

[c9af49f](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/c9af49f5bc92d7927d000e1f8a0a7015d1a0e61d)

RonTuretzky commented on src/contracts/SavingCircles.sol:
```diff
@@ -191,7 +178,7 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
     isMember[_id][msg.sender] = true;
     memberCircles[msg.sender].push(_id);
-    _circle.members.push(msg.sender);
+    circleMembers[_id].push(msg.sender);
```
I dont see a max count validation here? I guess we are just trusting that the owner issues a finite amount of members but maybe we should add a comment to this

exo404 replied:
The only other option I can think of is adding a size field to the saving circle struct but that would be used only for this and maybe will not reflect the effective number of members so I guess a comment is better

exo404 replied:
I think the frontend should generate all the invites at once, after the owner gives the max size input. We can make a validation there checking if the size is less than a constant MAX (maybe 20/25 like the optimal size for safety nets)

exo404 replied:
[57afc54](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/57afc54096c6e1bdb5cbfd6f7d81d34649bdfca8)

RonTuretzky reviewed (changes requested):
recent changes look good, small nit and copilot comments

RonTuretzky commented on app/invites.tsx:
```diff
@@ -0,0 +1,606 @@
+  'function redeemInvite(uint256 id, uint256 nonce, bytes signature)',
+  'function isMember(uint256 id, address member) view returns (bool)',
+  'function owner() view returns (address)',
+  'function getCircle(uint256 id) view returns (address owner, address[] members, uint256 currentIndex, uint256 depositAmount, address token, uint256 depositInterval, uint256 circleStart, uint256 circleEnd)'
```
@exo404

exo404 replied:
[f0fee14](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/f0fee142b2f8543d7292d2d0438082e0e2c5b281)

exo404 committed:
refactor: add test shared functions

exo404 committed:
refactor: fuzz test to use utils

exo404 committed:
refactor: unit test to use utils

exo404 committed:
refactor: invariant test to use utils

exo404 committed:
refactor: integration test to use utils

exo404 committed:
fix: getCircle ABI

exo404 commented on test/unit/SavingCirclesUnit.t.sol:
```diff
@@ -64,19 +74,111 @@ contract SavingCirclesUnit is Test {
+      savingCircles.redeemInvite(_circleId, nonce, signature);
+      nonce++;
+    }
+  }
```
[5201e5f](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/5201e5fbf12872dd19f8a2d163058a14d88b8569)

[bf432fe](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/bf432fe579f5f5734fcebbd813e1cc604b1bf2ea)

[ed893b6](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/ed893b6a4e8870d956e3a4c7caae35494fc1aedc)

[6aeb0d1](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/6aeb0d125668a8b3b7f18c21c65a3248b09f6533)

[c9af49f](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/c9af49f5bc92d7927d000e1f8a0a7015d1a0e61d)

exo404 committed:
chore: adding max members validation comment

exo404 committed:
feat: issue #68

exo404 committed:
feat: add isDecommisionable to viewer contract (issue #68)

exo404 commented:
@franrolotti I implemented the changes for the issue #68 in these two commits: [0463444](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/04634440aaa34910bf08eadba01f3f1dcd8fb1fb) [e283738](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/e283738677d4313a2cf91b1698d7330f942928fe)

RonTuretzky commented:
Check warning on line R148
Check warning:
Possible reentrancy vulnerabilities. Avoid state changes after transfer
@exo404

bagelface commented on src/contracts/SavingCircles.sol:
```diff
@@ -63,80 +67,113 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+
   /// @inheritdoc ISavingCircles
   function deposit(uint256 _id, uint256 _value) external override nonReentrant {
+    if (!isActive[_id]) revert NotActive();
```
could probably make this into an "onlyActive(uint256 _id)" modifier

exo404 replied:
[18cddc8](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/18cddc863a42c878fd8924d9615c2683dd45a27e)

bagelface commented on src/contracts/SavingCircles.sol:
```diff
@@ -179,53 +216,67 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+
   /// @inheritdoc ISavingCircles
   function isWithdrawable(uint256 _id) public view override returns (bool) {
+    if (!isActive[_id]) revert NotActive();
```
this should just return false if it's not active, rather than reverting.

exo404 replied:
[18cddc8](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/18cddc863a42c878fd8924d9615c2683dd45a27e)

bagelface commented on src/contracts/SavingCircles.sol:
```diff
@@ -179,53 +216,67 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
 
   /// @inheritdoc ISavingCircles
   function withdrawableBy(uint256 _id) public view override onlyCommissioned(_id) returns (address) {
+    if (!isActive[_id]) revert NotActive();
```
this should just return `address(0)` if it's not active, rather than reverting.

exo404 replied:
[18cddc8](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/18cddc863a42c878fd8924d9615c2683dd45a27e)

bagelface reviewed (commented):
A few small issues and I think we should address the Github generated warnings.

One thing I think we should should consider is that since we don't specify addresses in the signature, the invites can be front run (monitor mempool and send a malicious `redeemtInvite()` transaction that gets cleared before the honest transaction). I don't think this is something that there's any financial incentive to do, so it's not a huge deal but there is potential for abuse here.

bagelface commented on src/contracts/SavingCircles.sol:
```diff
@@ -295,4 +348,50 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+   * @notice _eip712DomainTypehash is keccak256('EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)')
+   * @notice Domain name is 'StacksInvite' and version is '1'
+   */
+  function _hashInvite(uint256 _id, uint256 _nonce) private view returns (bytes32) {
```
I don't love the use of "magic numbers" here. Typically, this is done with compile time constants. It's more readable and probably slightly more efficient. I also think we should be implementing OpenZeppelin's `EIP712Upgradeable`.

exo404 replied:
about the magic numbers, first version of the function used constants in the contract but @RonTuretzky said that he prefers to put only the hashes. I personally hate magic numbers too

exo404 replied:
I'll implement that interface to address this issue

exo404 replied:
[3312a1e](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/3312a1e9450f8c1ad97df79d08a4cd438a78c4c6)

exo404 commented:
I'll try to address Github linter warnings and warnings in the code (reentracncy pinted by @RonTuretzky) plus the few changes requested by @bagelface in a few hours

exo404 committed:
feat: add onlyActive modifier

exo404 committed:
fix: reentrancy fixed

exo404 commented:
> Check warning on line R148 Check warning: Possible reentrancy vulnerabilities. Avoid state changes after transfer @exo404

[faead74](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/faead74e07b9fe33737b006f151c1c2f3ac9c8e9)

exo404 committed:
fix: fix remappings and add EIP712Upgradable interface to the contract

exo404 committed:
fix: addressing linter warnings

exo404 commented:
this is for github warnings @bagelface [c7388c4](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/c7388c4b2fd80803d7cfbe5f1abf450d3df4942e)

exo404 committed:
fix: gas improvements and natspec

exo404 commented:
[3312a1e](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/3312a1e9450f8c1ad97df79d08a4cd438a78c4c6) @franrolotti

exo404 committed:
fix: linter warnings

exo404 commented:
@bagelface last warnings here [fix: linter warnings](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/79f01acfec2d971c4a8b212723a21ffe22b31589)

bagelface commented on src/contracts/SavingCircles.sol:
```diff
@@ -54,6 +58,7 @@ contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
   /// @inheritdoc ISavingCircles
   function initialize(address _owner) external override initializer {
+    __EIP712_init(EIP712_NAME, EIP712_VERSION);
     __Ownable_init_unchained(_owner);
```
Sorry, just notice this. Why are we using the "unchained" version here? Should probably just call `__Ownable_init(_owner)`.

exo404 replied:
[714e277](https://github.com/BreadchainCoop/saving-circles/pull/52/commits/714e2779c7c1b322efeb52a077d87f522e2f5f98)

bagelface reviewed (approved):
one last small comment then it looks good to me

exo404 committed:
fix: using the correct init function for Ownable interface
